pkg/controller/workload/mpijob/mpijob_controller.go

```diff
@@ -0,0 +1,579 @@
+		var w *kueue.Workload
+		if len(workloads.Items) == 1 {
+			// The job may have been modified and hence the existing workload
+			// doesn't match the job anymore. All bets are off if there are more
+			// than one workload...
+			w = &workloads.Items[0]
+		}
+		if err := r.stopJob(ctx, w, job, "No matching Workload"); err != nil {
+			log.Error(err, "stopping job")
+		}
+	}
+
+	// Delete duplicate workload instances.
+	existedWls := 0
+	for i := range toDelete {
+		err := r.client.Delete(ctx, toDelete[i])
+		if err == nil || !apierrors.IsNotFound(err) {
+			existedWls++
+		}
+		if err != nil && !apierrors.IsNotFound(err) {
+			log.Error(err, "Failed to delete workload")
+		}
+		if err == nil {
+			r.record.Eventf(job, corev1.EventTypeNormal, "DeletedWorkload",
+				"Deleted not matching Workload: %v", workload.Key(toDelete[i]))
+		}
+	}
+
+	if existedWls != 0 {
+		if match == nil {
+			return nil, fmt.Errorf("no matching workload was found, tried deleting %d existing workload(s)", existedWls)
+		}
+		return nil, fmt.Errorf("only one workload should exist, found %d", len(workloads.Items))
+	}
+
+	return match, nil
+}
+
+func ConstructWorkloadFor(ctx context.Context, client client.Client,
+	job *kubeflow.MPIJob, scheme *runtime.Scheme) (*kueue.Workload, error) {
+	w := &kueue.Workload{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      GetWorkloadNameForMPIJob(job.Name),
+			Namespace: job.Namespace,
+		},
+		Spec: kueue.WorkloadSpec{
+			QueueName: queueName(job),
+		},
+	}
+
+	highestPriorityClassName := ""
+	highestPriority := int32(0)
+
+	for index, mpiReplicaType := range orderedReplicaTypes(&job.Spec) {
+		podSet := kueue.PodSet{
+			Name:  strconv.Itoa(index),
+			Spec:  *job.Spec.MPIReplicaSpecs[mpiReplicaType].Template.Spec.DeepCopy(),
+			Count: podsCount(&job.Spec, mpiReplicaType),
+		}
+		w.Spec.PodSets = append(w.Spec.PodSets, podSet)
+
+		priorityClassName, p, err := utilpriority.GetPriorityFromPriorityClass(
+			ctx, client, job.Spec.MPIReplicaSpecs[mpiReplicaType].Template.Spec.PriorityClassName)
+		if err != nil {
+			return nil, err
+		}
+		if p > highestPriority {
+			highestPriority = p
+			highestPriorityClassName = priorityClassName
+		}
+	}
+	// Populate priority from priority class.
+	w.Spec.Priority = &highestPriority
+	w.Spec.PriorityClassName = highestPriorityClassName
+
+	if err := ctrl.SetControllerReference(job, w, scheme); err != nil {
+		return nil, err
+	}
+
+	return w, nil
+}
+
+func orderedReplicaTypes(jobSpec *kubeflow.MPIJobSpec) []kubeflow.MPIReplicaType {
+	var result []kubeflow.MPIReplicaType
+	for mpiReplicaType := range jobSpec.MPIReplicaSpecs {
+		result = append(result, mpiReplicaType)
+	}
+	sort.Slice(result, func(i, j int) bool {
+		return result[i] < result[j]
+	})
+	return result
+}
+
+func podsCount(jobSpec *kubeflow.MPIJobSpec, mpiReplicaType kubeflow.MPIReplicaType) int32 {
+	return pointer.Int32Deref(jobSpec.MPIReplicaSpecs[mpiReplicaType].Replicas, 1)
```

Comment:
https://github.com/kubeflow/mpi-operator/blob/b302019be7af803ab4aa15e99d4e5c5b4550e357/pkg/apis/kubeflow/validation/validation.go#L128

replicas can actually be zero for worker. Note that mpi-operator doesn't include a webhook, so you can't assume the object is validated here.

Reply:
Indeed, this line only does fallback to 1, following the documentation for the field: https://github.com/kubeflow/common/blob/4dbf18c85417b54158914d1d01c1744b4e3542f2/pkg/apis/common/v1/types.go#L80.

Reply:
> Indeed, this line only does fallback to 1, following the documentation for the field:

IIUC, the comment is inaccurate for MPIJob

Reply:
Indeed, MPIJob validation fails with: `Found validation errors: spec.mpiReplicaSpecs[Worker].replicas: Invalid value: 0: must be greater than or equal to 1`. The workload is never created in that case. 

Still, the fallback makes sense as this seems to be an issue with MPIJob, and will work well once fixed there.

Reply:
Ah, thanks for checking. We can proceed with the defaulting here then.